### DISCUSSION TITLE
How to set page referrer manually?

### DISCUSSION THREAD

**Discussion by vivek-yes** [vercel/next.js#60546] [App Router]
Is there a way to manually set Previous page url for page referrer when we navigate to each page on next.js

we wanted the previous page url to be set for page referrer when we navigate to each page.

**Comment by icyJoseph** [comment]
This is in general a problem with Single Page Applications, or rather, more technically, a problem with apps that use the history API to navigate without unloading and reloading the page at the new URL location. 

With Pages Router, you could listen to Routing events, and with that update your trackers and such, to tell them where the navigation came from, within your application. 

- https://mbaersch.medium.com/attribution-problems-in-spas-solutions-and-caveats-9362fbe8ac3a
- https://help.adtriba.com/en/articles/3733261-how-to-deal-with-single-page-applications
- https://www.simoahava.com/gtm-tips/fix-rogue-referral-problem-single-page-sites/
- https://www.linkedin.com/pulse/how-fix-ga4-rogue-referral-problem-spa-websites-ortolani-barbosa

In App Router, there's also a way to construct a hook that'll let you know about routing events:

- https://nextjs.org/docs/app/api-reference/functions/use-router#router-events

And with that update the data you sent to your tracker service. 

Do remember that the `document.referrer` property, is read only.

**Comment by vivek-yes** [comment]
Hi icyJoseph, Thanks for the reply...

Basically I wanted to update the page referrer on google analytics. 

![pageviews](https://github.com/vercel/next.js/assets/36656693/8077d32e-75a1-41a5-8c99-dc3f6aa539a2)

When ever we navigate through some page we wanted the page referrer value as previous page on google analytics. Could you recommend a solution for this

**Comment by ScottEAdams** [comment]
There is actually a sneaky workaround for this using the ServerTiming api where you do something like this in middleware
```
// Set a new response header `server-timing`
const requestHeaders = new Headers(request.headers)
response.headers.set('server-timing', `true-referer;desc="${requestHeaders.get('referer')}"`)
return response
```

Which can then be accessed in the client via
```
if (window.performance && performance.getEntriesByType) {
	const navTiming = performance.getEntriesByType('navigation')
	if (navTiming.length > 0) {
		const serverTiming = navTiming[0].serverTiming
		if (serverTiming && serverTiming.length > 0) {
			for (const element of serverTiming) {
				console.log(`${element.name} = ${element.description}`)
			}
		}
	}
}
```
